From a188445accb5db9414599ad350ded195047a6661 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Wed, 6 Sep 2023 21:25:55 -0500
Subject: [PATCH 1/2] implement regex support in rules for app_id and title

Enables the use of regular expressions for window rules "app_id" and "title"
---
 config.def.h |  1 +
 dwl.c        | 19 +++++++++++++++++--
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/config.def.h b/config.def.h
index 1677f6f12..1eb646127 100644
--- a/config.def.h
+++ b/config.def.h
@@ -17,6 +17,7 @@ static const Rule rules[] = {
 	{ "Gimp",     NULL,       0,            1,           -1 },
 	*/
 	{ "firefox",  NULL,       1 << 8,       0,           -1 },
+	{ "^kitty$",  NULL,       0,            0,           -1 },
 };
 
 /* layout(s) */
diff --git a/dwl.c b/dwl.c
index 4ff5c37f7..9c29eb4aa 100644
--- a/dwl.c
+++ b/dwl.c
@@ -11,6 +11,7 @@
 #include <sys/wait.h>
 #include <time.h>
 #include <unistd.h>
+#include <regex.h>
 #include <wayland-server-core.h>
 #include <wlr/backend.h>
 #include <wlr/backend/libinput.h>
@@ -318,6 +319,7 @@ static Monitor *xytomon(double x, double y);
 static void xytonode(double x, double y, struct wlr_surface **psurface,
 		Client **pc, LayerSurface **pl, double *nx, double *ny);
 static void zoom(const Arg *arg);
+static int regex_match(const char *pattern, const char *str);
 
 /* variables */
 static const char broken[] = "broken";
@@ -457,8 +459,8 @@ applyrules(Client *c)
 		title = broken;
 
 	for (r = rules; r < END(rules); r++) {
-		if ((!r->title || strstr(title, r->title))
-				&& (!r->id || strstr(appid, r->id))) {
+		if ((!r->title || regex_match(title, r->title))
+				&& (!r->id || regex_match(appid, r->id))) {
 			c->isfloating = r->isfloating;
 			newtags |= r->tags;
 			i = 0;
@@ -2683,6 +2685,19 @@ zoom(const Arg *arg)
 	arrange(selmon);
 }
 
+int
+regex_match(const char *pattern, const char *str) {
+  regex_t regex;
+  int reti;
+  if (regcomp(&regex, pattern, REG_EXTENDED) != 0)
+    return 0;
+  reti = regexec(&regex, str, (size_t)0, NULL, 0);
+  regfree(&regex);
+  if (reti == 0)
+    return 1;
+  return 0;
+}
+
 #ifdef XWAYLAND
 void
 activatex11(struct wl_listener *listener, void *data)

From d7160c26d3f0f131a6966907fef7e926c6eba781 Mon Sep 17 00:00:00 2001
From: wochap <gean.marroquin@gmail.com>
Date: Wed, 11 Oct 2023 23:09:45 -0500
Subject: [PATCH 2/2] fix typo

---
 dwl.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dwl.c b/dwl.c
index 9c29eb4aa..41ec97ba6 100644
--- a/dwl.c
+++ b/dwl.c
@@ -459,8 +459,8 @@ applyrules(Client *c)
 		title = broken;
 
 	for (r = rules; r < END(rules); r++) {
-		if ((!r->title || regex_match(title, r->title))
-				&& (!r->id || regex_match(appid, r->id))) {
+		if ((!r->title || regex_match(r->title, title))
+				&& (!r->id || regex_match(r->id, appid))) {
 			c->isfloating = r->isfloating;
 			newtags |= r->tags;
 			i = 0;
